File: benchmarks/writer_perf.c
// Writer Protocol Performance Benchmark - C baseline
// Measures direct-to-buffer formatting using byte-by-byte loops
// Matches Salt's current loop-based write_str implementation for fair
// comparison

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define ITERATIONS 10000000
#define BUFFER_SIZE 4096

// Simulate a "Writer" that accumulates to a buffer
typedef struct {
  char *data;
  size_t len;
  size_t capacity;
} Buffer;

// Byte-by-byte push (matches Salt's push_byte approach)
static inline void buffer_push_byte(Buffer *b, char c) {
  if (b->len < b->capacity) {
    b->data[b->len++] = c;
  }
}

// Loop-based write_str (matches Salt's current approach)
static inline void buffer_write_str(Buffer *b, const char *s, size_t slen) {
  for (size_t i = 0; i < slen; i++) {
    buffer_push_byte(b, s[i]);
  }
}

// Inlined integer formatting (digit extraction + byte-by-byte push)
static inline void buffer_write_i32(Buffer *b, int val) {
  char tmp[16];
  int len = 0;
  int n = val;
  int neg = n < 0;
  if (neg)
    n = -n;

  if (n == 0) {
    buffer_push_byte(b, '0');
    return;
  }

  while (n > 0) {
    tmp[len++] = '0' + (n % 10);
    n /= 10;
  }

  if (neg)
    buffer_push_byte(b, '-');

  for (int i = len - 1; i >= 0; i--) {
    buffer_push_byte(b, tmp[i]);
  }
}

// Inlined i64 formatting (matches Salt's write_i64)
static inline void buffer_write_i64(Buffer *b, long val) {
  char tmp[24];
  int len = 0;
  long n = val;
  int neg = n < 0;
  if (neg)
    n = -n;

  if (n == 0) {
    buffer_push_byte(b, '0');
    return;
  }

  while (n > 0) {
    tmp[len++] = '0' + (n % 10);
    n /= 10;
  }

  if (neg)
    buffer_push_byte(b, '-');

  for (int i = len - 1; i >= 0; i--) {
    buffer_push_byte(b, tmp[i]);
  }
}

int main(int argc, char **argv) {
  char storage[BUFFER_SIZE];
  Buffer buf = {.data = storage, .len = 0, .capacity = BUFFER_SIZE};

  long total_len = 0;

  for (int i = 0; i < ITERATIONS; i++) {
    // Reset buffer each iteration
    buf.len = 0;

    // Direct write calls using byte-by-byte approach
    buffer_write_str(&buf, "Item ", 5);
    buffer_write_i32(&buf, i);
    buffer_write_str(&buf, ": val = ", 8);
    buffer_write_i64(&buf, (long)i * 1000);

    total_len += buf.len;
  }

  // Prevent DCE
  if (total_len == 0)
    return 1;
  return 0;
}

File: benchmarks/writer_perf.rs
// Writer Protocol Performance Benchmark - Rust
// Measures direct-to-buffer formatting using byte-by-byte loops
// Matches Salt's current loop-based write_str implementation for fair comparison

const ITERATIONS: usize = 10_000_000;
const BUFFER_SIZE: usize = 4096;

struct Buffer {
    data: Vec<u8>,
    len: usize,
}

impl Buffer {
    fn new(capacity: usize) -> Self {
        Buffer {
            data: vec![0u8; capacity],
            len: 0,
        }
    }
    
    fn clear(&mut self) {
        self.len = 0;
    }
    
    // Byte-by-byte push (matches Salt's push_byte)
    fn push_byte(&mut self, b: u8) {
        if self.len < self.data.len() {
            self.data[self.len] = b;
            self.len += 1;
        }
    }
    
    // Loop-based write_str (matches Salt's approach)
    fn write_str(&mut self, s: &[u8]) {
        for &b in s {
            self.push_byte(b);
        }
    }
    
    // Integer formatting with byte-by-byte push
    fn write_i32(&mut self, val: i32) {
        if val == 0 {
            self.push_byte(b'0');
            return;
        }
        
        let mut tmp = [0u8; 12];
        let mut count = 0;
        let mut n = val;
        let neg = n < 0;
        if neg { n = -n; }
        
        while n > 0 {
            tmp[count] = b'0' + (n % 10) as u8;
            count += 1;
            n /= 10;
        }
        
        if neg { self.push_byte(b'-'); }
        
        for i in (0..count).rev() {
            self.push_byte(tmp[i]);
        }
    }
    
    fn write_i64(&mut self, val: i64) {
        if val == 0 {
            self.push_byte(b'0');
            return;
        }
        
        let mut tmp = [0u8; 20];
        let mut count = 0;
        let mut n = val;
        let neg = n < 0;
        if neg { n = -n; }
        
        while n > 0 {
            tmp[count] = b'0' + (n % 10) as u8;
            count += 1;
            n /= 10;
        }
        
        if neg { self.push_byte(b'-'); }
        
        for i in (0..count).rev() {
            self.push_byte(tmp[i]);
        }
    }
    
    fn len(&self) -> usize {
        self.len
    }
}

fn main() {
    let mut buf = Buffer::new(BUFFER_SIZE);
    let mut total_len: i64 = 0;
    let mut checksum: u8 = 0;
    
    for i in 0..ITERATIONS {
        buf.clear();
        
        // Direct write calls using byte-by-byte approach
        buf.write_str(b"Item ");
        buf.write_i32(i as i32);
        buf.write_str(b": val = ");
        buf.write_i64((i as i64) * 1000);
        
        total_len += buf.len() as i64;
        
        // [FAIR COMPARISON] XOR first AND last byte to prevent Dead Store Elimination
        // This creates a data dependency forcing ALL buffer writes including integer formatting
        checksum ^= buf.data[0];
        checksum ^= buf.data[buf.len - 1];
    }
    
    // Prevent DCE - use both length and checksum
    if total_len == 0 || checksum == 255 {
        std::process::exit(1);
    }
}

